Moving streaming library to staging/cri-api/pkg/streaming

alok87 opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This PR is required to make life easier for people developing on cri-api, so that they do not need to vendor kubernetes for this library (as vendoring k8s is hard)
- [x] Move out streaming library to `staging/cri-api/pkg/streaming`
- [x] Isolate `remotecommand` and `portforward` dependencies
- [x] Isolate `apis/core`

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/75828

**KEP under discussion**:
https://github.com/kubernetes/enhancements/pull/937

**Special notes for your reviewer**:
None
cc @dims 

**Does this PR introduce a user-facing change?**:
None

```release-note
Streaming library has a new location under staging/cri-api/pkg/streaming
```

Commit:
Streaming api moved to a new home under staging/cri-api/pkg

Refer to https://github.com/kubernetes/kubernetes/issues/75828#issue-426475237

Moving streaming under staging/cri-api/pkg as this library is indenpdent
to cri-api and is more of a helper library that can be reused for a
different use case as well.

alok87 commented:
/hold

alok87 commented:
/unhold

alok87 commented:
/ok-to-test

alok87 commented:
/hold cancel

dims commented:
/ok-to-test

alok87 commented:
/retest

nikhita commented on staging/src/k8s.io/cri-api/go.mod:
```diff
@@ -7,17 +7,15 @@ module k8s.io/cri-api
+	k8s.io/apimachinery v0.0.0
+	k8s.io/apiserver v0.0.0 // indirect
+	k8s.io/client-go v0.0.0
+	k8s.io/kubernetes v1.14.0
```
staging repos shouldn't be using k/k

nikhita replied:
https://github.com/kubernetes/kubernetes/blob/97011635f727307db25566a0459a51c6e1663836/staging/src/k8s.io/cri-api/pkg/streaming/server.go#L38

liggitt replied:
also:
https://github.com/kubernetes/kubernetes/blob/97011635f727307db25566a0459a51c6e1663836/staging/src/k8s.io/cri-api/pkg/streaming/server.go#L39

liggitt commented on staging/src/k8s.io/cri-api/go.mod:
```diff
@@ -7,17 +7,15 @@ module k8s.io/cri-api
+	k8s.io/apimachinery v0.0.0
+	k8s.io/apiserver v0.0.0 // indirect
+	k8s.io/client-go v0.0.0
+	k8s.io/kubernetes v1.14.0
```
things in staging cannot depend on `k8s.io/kubernetes`

liggitt commented:
looking through this, do we want the api and streaming implementation combined into a single module? the added import tree is significant for someone who just wanted the proto definitions

Commit:
Isolating portforward and remotecommand

Nothing in staging can import k8s.io/kubernetes so isolating these
dependencies by moving them to cri-api/pkg/streaming

alok87 commented:
@nikhita @liggitt thanks for the review.

Opening a KEP for figuring out the location for KEP.

I saw few more dependencies after I moved `portforward` and `remotecommand`. 
```
./pkg/streaming/server_test.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/httpstream.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/httpstream_test.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/websocket.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/remotecommand/httpstream.go:	api "k8s.io/kubernetes/pkg/apis/core"
```

Any suggestion on isolating these dependencies? @dims @nikhita @liggitt

Commit:
Isolating pkg/apis/core with "k8s.io/api/core/v1"

caesarxuchao commented:
/assign @roycaihw

alok87 commented:
/hold 
a KEP is being raise to discuss on where should be the location of streaming library
- `staging/cri-api/pkg/streaming`
- `staging/cri-server`
- `staging/cri-helpers` or with some other name

roycaihw commented:
/remove-sig api-machinery
/unassign

nikhita commented on staging/publishing/rules.yaml:
```diff
@@ -322,4 +322,13 @@ rules:
   - source:
       branch: master
       dir: staging/src/k8s.io/cri-api
+    dependencies:
```
Can you move `name: master` above dependencies?

Sorry for not pointing this out earlier, but you could also add import restrictions for cri-api here: https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/import-restrictions.yaml? Thanks!

alok87 replied:
Move the `name: master`
Added import restrictions based on https://github.com/alok87/kubernetes/blob/dep-streaming-moved-out/staging/src/k8s.io/cri-api/go.mod#L15-L20

Why do we specify import restrictions? What does the bot do reading this file?

nikhita replied:
> Added import restrictions based on https://github.com/alok87/kubernetes/blob/dep-streaming-moved-out/staging/src/k8s.io/cri-api/go.mod#L15-L20

Ah, I meant adding the import restrictions to this file: https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/import-restrictions.yaml.

> Why do we specify import restrictions? What does the bot do reading this file?

It makes sure that a package won't import any additional packages accidentally. Importing a new package would mean explicitly adding it in the list in that file.

The main significance of that file is to make sure that a staging repo won't import k/k though...

alok87 replied:
Got it. thanks
I have added it in `import-restrictions.yaml` only

nikhita replied:
@alok87 I can't see the change in the diff. Maybe forgot to push? :upside_down_face:

alok87 replied:
please check now :)

alok87 commented:
/retest

alok87 commented:
/retest

alok87 commented:
@nikhita Can you help me figure out the reason for bazel verification failure. Update bazel script is working fine locally. Also what is bazel build, any doc link could help.

nikhita commented:
@alok87 can you rebase and retry using `hack/update-bazel.sh`? If that doesn't work, let us know :)

dims commented:
/uncc

spiffxp commented:
/remove-area conformance

dims commented:
/area code-organization

dims commented:
/close

let's revisit this after the KEP merges
